### PR TITLE
Added new command to list, add & delete build tags

### LINKED ISSUE #215: Feature Request: Add build tag via cli
In addition of creating builds via CLI, it will indeed be great if we can update build tags while creating or after the fact it's created

### PR DESCRIPTION
**Pull request by dhilmathy**
Fixes #215

### CODE REVIEW

**Comment by atbagga** [comment]
@dhilmathy Apologies for long delays. 

We have converted vsts-cli to Azure Devops Extension, which is an extension to Azure Cli. Refer repository readme for more details.  

Would you like to contribute this PR by porting it to the extension code. It should not be major changes for porting.

File: azure-devops/azext_devops/dev/pipelines/build.py [hunk 1]
```diff
@@ -151,6 +151,86 @@ def build_list(definition_ids=None, branch=None, organization=None, project=None
...
+            tags = client.add_build_tag(project=project, build_id=build_id, tag=tags[0])
+        else:
+            tags = client.add_build_tags(tags=tags, project=project, build_id=build_id)
+        return tags
+    except VstsServiceError as ex:
+        raise CLIError(ex)
+
+
+def delete_build_tag(build_id, tag, organization=None, project=None, detect=None):
+    """Deletes a tag from a build.
```
Comment (atbagga) [inline azure-devops/azext_devops/dev/pipelines/build.py]: nit: Delete a build tag.

File: azure-devops/azext_devops/dev/pipelines/build.py [hunk 2]
```diff
@@ -151,6 +151,86 @@ def build_list(definition_ids=None, branch=None, organization=None, project=None
...
+                                                             project=project)
+        client = get_build_client(organization)
+        tags = client.delete_build_tag(project=project, build_id=build_id, tag=tag)
+        return tags
+    except VstsServiceError as ex:
+        raise CLIError(ex)
+
+
+def get_build_tags(build_id, organization=None, project=None, detect=None):
+    """Gets the tags for a build
```
Comment (atbagga) [inline azure-devops/azext_devops/dev/pipelines/build.py]: nit: Get tags for a build.

File: azure-devops/azext_devops/dev/pipelines/build.py [hunk 3]
```diff
@@ -151,6 +151,86 @@ def build_list(definition_ids=None, branch=None, organization=None, project=None
...
+    :param detect: Automatically detect instance and project. Default is "on".
+    :type detect: str
+    :rtype: list of str
+    """
+    try:
+        organization, project = resolve_instance_and_project(detect=detect,
+                                                             organization=organization,
+                                                             project=project)
+        client = get_build_client(organization)
+        if build_id is None:
```
Comment (atbagga) [inline azure-devops/azext_devops/dev/pipelines/build.py]: Will never reach here. BuildId is compulsory parameter.

File: azure-devops/azext_devops/dev/pipelines/commands.py [hunk 4]
```diff
@@ -44,6 +44,12 @@ def load_build_commands(self, _):
         g.command('queue', 'build_queue', table_transformer=transform_build_table_output)
         g.command('show', 'build_show', table_transformer=transform_build_table_output)
 
+    with self.command_group('pipelines build tags', command_type=buildOps) as g:
+        # basic build tag commands
+        g.command('list', 'get_build_tags', table_transformer=transform_tasks_table_output)
```
Comment (atbagga) [inline azure-devops/azext_devops/dev/pipelines/commands.py]: Will the task table output work? I think you will need to define the table format for this..

File: azure-devops/azext_devops/dev/pipelines/_help.py [hunk 5]
```diff
@@ -19,6 +19,12 @@ def load_pipelines_help():
     long-summary:
     """
 
+    helps['pipelines build tags'] = """
```
Comment (atbagga) [inline azure-devops/azext_devops/dev/pipelines/_help.py]: tag instead of tags

**Comment by dhilmathy** [comment]
@atbagga I'd added Unit Tests. As part of it, i am trying to queue a test build for `BuildTests Definition1` under `buildtests` project. But my test case execution failing with error : `TF400813: The user '' is not authorized to access this resource.` Any idea?

**Comment by atbagga** [comment]
> @atbagga I'd added Unit Tests. As part of it, i am trying to queue a test build for `BuildTests Definition1` under `buildtests` project. But my test case execution failing with error : `TF400813: The user '' is not authorized to access this resource.` Any idea?

The test organization is set as https://dev.azure.com/azuredevopsclitest which you won't have accesss to.  we are fixing the tests to allow external users to add recordings easily.  

Check this.. 
https://github.com/Microsoft/azure-devops-cli-extension/blob/b7c8d8e91e0ee4b774e970164a50863a6bb8bf09/tests/test_reposPoliciesTest.py

DEVOPS_CLI_TEST_ORGANIZATION = get_test_org_from_env_variable() or 'Https://dev.azure.com/baggaatul24'

Test is recorded against my personal account here.. 'Https://dev.azure.com/baggaatul24'

Since this get_test_org_from_env_variable()  is not available in your code you can avoid importing the constant DEVOPS_CLI_TEST_ORGANIZATION in your test and directly assign it to your account (for which you have valid PAT token) 

You will need to add the PAT to environment variable (AZURE_DEVOPS_EXT_PAT) so the test can access it.

**Comment by atbagga** [comment]
> @atbagga sorry, i forgot to mention that the error which i was getting is when i pushed my changes.
> 
> **Azure DevOps CLI - PR GitHub YAML** (_#20190131.1_) is failing with that error. Surprising part for me, queuing a _test build_ is working in existing test `test_pipelinesBuildTest`. But, it's failing under `test_pipelinesBuildTagTest` (the one i added for build tag).
> 
> Thanks for your detailed information on how to run it locally.

Ahh ok.. Every test has a recording file which should be submitted with the test. In the pipeline the tests will run depending on recording for API calls. In case recording is missing it tries to make an actual network call (this will fail because the pat token in the tests is dummy token) 

In the CI run.. your test does not have a recording in the folder here - https://github.com/Microsoft/azure-devops-cli-extension/tree/master/tests/recordings

The other test does so that succeeds and yours fails.

You can try the steps recommended in above comment for getting a recording file for your test. 
I also updated test authoring guidelines with the same. 

Do let us know if you face any issues in trying it out.

**Comment by atbagga** [comment]
@dhilmathy  Delete the existing test recording. Clear the python client cache (.vsts folder in your homepath). 
Run this test individually e.g. 
```pytest ./tests/test_pipelinesBuildTagTest.py```

The issue is the recording does not have the options call which is cached in .vsts folder. So the test might succeed in live mode but will fail in recording. 

I will update this is docs and try to add some script probably so others don't got through the pain. 
Feel free to suggest changes in the docs to improve the experience for new contributors.

**Comment by atbagga** [comment]
@dhilmathy  Raised a PR in your repo with the recording. https://github.com/dhilmathy/azure-devops-cli-extension/pull/1
This should give you access to the recording with the test account. 

We have an open issue to make test authoring easy. I am taking note of the issues you faced. Hopefully we will improve the experience soon.

**Pull request merged**